File: src/Pointer.ts
import {Path} from './Path';
import {Container} from './Container';
import {InkObject} from './Object';

export class Pointer{
  public container: Container | null;
  public index: number;

	constructor(container: Container | null, index: number){
		this.container = container;
		this.index = index;
	}

	public Resolve(): Container | InkObject | null{
		if (this.index < 0) return this.container;
		if (this.container == null) return null;
		if (this.container.content.length == 0) return this.container;
		if (this.index >= this.container.content.length) return null;

		return this.container.content[this.index];
	}

	get isNull(): boolean{
		return this.container == null;
	}

	get path(): Path | null{
		if (this.isNull) return null;

		if (this.index >= 0)
			return this.container!.path.PathByAppendingComponent(new Path.Component(this.index));
		else
			return this.container!.path;
	}

	public toString(): string{
		if (!this.container)
			return 'Ink Pointer (null)';

		return 'Ink Pointer -> ' + this.container.path.toString() + ' -- index ' + this.index;
	}

	// This method does not exist in the original C# code, but is here to maintain the
	// value semantics of Pointer.
	public copy(): Pointer{
		return new Pointer(this.container, this.index);
	}

	public static StartOf(container: Container | null): Pointer{
		return new Pointer(container, 0);
	}

	public static get Null(): Pointer {
		return new Pointer(null, -1);
	}
}

File: ink_dart/lib/src/engine/pointer.dart
import 'ink_object.dart';
import 'container.dart';
import 'path.dart';

/// A pointer to a specific position within a Container.
///
/// Combines a Container reference with an index into its content.
class Pointer {
  /// The container this pointer points into.
  Container? container;

  /// The index within the container's content.
  int index;

  Pointer(this.container, this.index);

  /// Create a pointer to the start of a container.
  Pointer.startOf(Container container) : this(container, 0);

  /// Create a null pointer.
  Pointer.null_()
      : container = null,
        index = -1;

  /// Whether this is a null pointer.
  bool get isNull => container == null;

  /// The element at this pointer's position.
  InkObject? get element {
    if (container == null || index < 0) return null;
    if (index >= container!.content.length) return null;
    return container!.content[index];
  }

  /// Resolve the pointer to get the object at this position.
  InkObject? resolve() {
    if (index < 0) return container;
    return element;
  }

  /// The path to this pointer's position.
  Path? get path {
    if (isNull) return null;
    return container!.path.pathByAppendingComponent(
      PathComponent.fromIndex(index),
    );
  }

  /// Create a copy of this pointer.
  Pointer copy() => Pointer(container, index);

  @override
  String toString() {
    if (isNull) {
      return 'Pointer.null';
    }
    return 'Pointer(${path?.componentsString ?? "?"})';
  }

  @override
  bool operator ==(Object other) {
    if (other is Pointer) {
      return container == other.container && index == other.index;
    }
    return false;
  }

  @override
  int get hashCode => Object.hash(container, index);
}
